flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -1339,14 +1368,14 @@ class ScalarExprLowering {
 
       if (arg.passBy == PassBy::MutableBox) {
         auto mutableBox = genMutableBoxValue(*expr);
-        auto IRbox = Fortran::lower::getMutableIRBox(builder, loc, mutableBox);
-        caller.placeInput(arg, IRbox);
+        auto irBox = Fortran::lower::getMutableIRBox(builder, loc, mutableBox);
+        caller.placeInput(arg, irBox);
         // TODO: no need to add this to the list if intent(in)
         mutableModifiedByCall.emplace_back(std::move(mutableBox));
         continue;
       }
 
-      auto argRef = genExtAddr(*expr);
+      auto argRef = genExtArg(*expr, arg.passBy == PassBy::Box);
```

Comment:
> I think will need to add more logic here, even if the argument is not passed by fir.box, we will also have to somehow lower designators with slices to something we can do copy-in/copy-out with.
I'm not sure I follow what you're getting at with this. Do you mean we might want to make a temp and then embox that? At this point, we've hit an argument in a function that *must* be passed in a box, so there really isn't any ambiguity on the calling convention. Note that if the expression *requires* a temporary be created, that is already done by the expression lowering code itself. Here, lowering is just making sure that the value passed is emboxed, which is required by the call interface.

> Bonus being that we can only lower the designator expression once, in case it has side effects in the indexes/or index expression whose evaluation would changed after the call.
That's not valid Fortran, and we can blame any unexpected behavior the user experiences on the user themselves. :)

Reply:
The kind of code I have in mind is things like `foo( x(1:12:istep()))` in the code below:

```
subroutine foo(x)
    real :: x(*)
    x(1) = 42.
    x(2) = 43.
end subroutine

function istep()
    integer, save :: i = 2
    istep = i
    i = i + 1
end function

   real :: x(4)
   x = 0.
   call foo(x(1:4:istep()))
   print *, x
end
```

`x(1:12:istep())` is a non contiguous definable entity that must be passed to foo that expects contiguous entity (f77 interface), so we need to deploy some kind of copy-in/copy-out mechanism here.

gfortran, ifort, nvfortran, xlf and nagfor that I tested all print `42. 0. 43. 0.`, which is what I expect with this code. If we are not careful and re-evaluate `istep()` in the copy-out, we would print `42. 0. 0. 43.`.

We would need to:
- copy from `x(1:12:istep())` into a temp [no problem, we know how to do that]
- call foo(temp) [no problem, we know how to do that]
- copy back from the temp into `x(1:12:istep())`, and there, I think, lies the small difficulty 

Right now, we might only be able to generate the copy-back by re-evaluating `x(1:12:istep())`, which would not yield the intended result, since `istep` result would be different from when the copy-in was made (3 instead of 2).
I do not think the fact that `istep` result changes on each calls makes it illegal to use here. From might reading, it does not violate 10.1.4 point 2 that limits side effects (Do you see a restriction that would apply here ?). It would be our compiler implementation that makes it necessary to evaluate it twice in the expression lowering, not the user code.

So my point was to say we might want to  evaluate `x(1:12:istep())` to a fir.box, and drive the copy-in/copy-out with that. Of course, there might be more clever way to do that using CPR, I did not think this through.

Again, we have not implemented copy-in copy-out yet, so that is more of a forethoughts related to your change rather than a request for something to happen in  this PR.